x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -825,4 +843,128 @@ private BitSet gatherPreAnalysisMetrics(LogicalPlan plan, BitSet b) {
+                        newChildren.add(e);
+                        continue;
+                    }
+                }
+                newChildren.add(args.get(i));
+            }
+            return childrenChanged ? f.replaceChildren(newChildren) : f;
+        }
+
+        private static Expression processArithmeticOperation(EsqlArithmeticOperation o) {
+            Expression left = o.left();
+            Expression right = o.right();
+            if (left.resolved() == false || right.resolved() == false) {
+                return o;
+            }
+            List<Expression> newChildren = new ArrayList<>(2);
+            boolean childrenChanged = false;
+            DataType targetDataType = NULL;
+            Expression from = Literal.NULL;
+
+            if (left.dataType() == KEYWORD && left.foldable() && (right.dataType().isNumeric() || right.dataType() == DATETIME)) {
+                targetDataType = right.dataType();
+                from = left;
+            }
+            if (right.dataType() == KEYWORD && right.foldable() && (left.dataType().isNumeric() || left.dataType() == DATETIME)) {
+                targetDataType = left.dataType();
+                from = right;
+            }
+
+            Expression e = from == Literal.NULL ? from : castStringLiteral(from, targetDataType);
+
+            if (e != Literal.NULL) {
+                childrenChanged = true;
+                newChildren.add(from == left ? e : left);
+                newChildren.add(from == right ? e : right);
+            }
+
+            return childrenChanged ? o.replaceChildren(newChildren) : o;
+        }
+
+        private static List<String[]> getMetaData(ScalarFunction f) {
+            List<String[]> paramTypes = new ArrayList<>();
+            var constructors = f.getClass().getConstructors();
+            Constructor<?> constructor = constructors[0]; // no multiple c'tors supported
+            var params = constructor.getParameters();
+            for (int i = 1; i < params.length; i++) { // skipping 1st argument, the source
+                if (Configuration.class.isAssignableFrom(params[i].getType()) == false) {
+                    Param paramInfo = params[i].getAnnotation(Param.class);
+                    String[] type = paramInfo == null ? new String[] { "?" } : paramInfo.type();
+                    paramTypes.add(type);
+                }
+            }
+            return paramTypes;
+        }
+
+        private static DataType getTargetType(String[] names) {
+            List<DataType> types = new ArrayList<>();
+            for (String name : names) {
+                types.add(DataTypes.fromEs(name));
+            }
+            try {
+                types.sort((dt1, dt2) -> dataTypeCastingPriority.get(dt1).compareTo(dataTypeCastingPriority.get(dt2)));
+            } catch (Exception e) {
+                return NULL;
+            }
+            return types.size() == 0 || isString(types.get(0)) ? NULL : types.get(0);
```

Comment:
You might be better off doing `Arrays.stream(name).min(yourcomparator).orElse(DataTypes.NULL)`. No need to fully sort I think.